File: portfolio web/python/snake.py
import random
import time
import js

# Snake game in Python for Pyodide
class SnakeGame:
    def __init__(self, width, height, cell_size):
        self.width = width
        self.height = height
        self.cell_size = cell_size
        self.grid_size = width // cell_size
        
        # Initialize snake
        self.snake = [(10, 10), (9, 10), (8, 10)]
        self.direction = "right"
        self.food = self.generate_food()
        self.score = 0
        self.game_over = False
        
    def generate_food(self):
        while True:
            x = random.randint(0, self.grid_size - 1)
            y = random.randint(0, self.grid_size - 1)
            
            # Make sure food doesn't spawn on snake
            if (x, y) not in self.snake:
                return (x, y)
    
    def change_direction(self, new_direction):
        # Prevent 180-degree turns
        if (new_direction == "up" and self.direction != "down" or
            new_direction == "down" and self.direction != "up" or
            new_direction == "left" and self.direction != "right" or
            new_direction == "right" and self.direction != "left"):
            self.direction = new_direction
    
    def move(self):
        if self.game_over:
            return
        
        # Get current head position
        head_x, head_y = self.snake[0]
        
        # Calculate new head position
        if self.direction == "up":
            head_y -= 1
        elif self.direction == "down":
            head_y += 1
        elif self.direction == "left":
            head_x -= 1
        elif self.direction == "right":
            head_x += 1
        
        # Check for wall collision
        if (head_x < 0 or head_x >= self.grid_size or 
            head_y < 0 or head_y >= self.grid_size):
            self.game_over = True
            return
        
        # Check for self collision
        if (head_x, head_y) in self.snake:
            self.game_over = True
            return
        
        # Add new head
        self.snake.insert(0, (head_x, head_y))
        
        # Check if food eaten
        if (head_x, head_y) == self.food:
            self.score += 10
            self.food = self.generate_food()
        else:
            # Remove tail if no food eaten
            self.snake.pop()
    
    def draw(self, ctx):
        # Clear canvas
        ctx.fillStyle = "#1f2937"
        ctx.fillRect(0, 0, self.width, self.height)
        
        # Draw snake
        for i, (x, y) in enumerate(self.snake):
            ctx.fillStyle = "#3b82f6" if i == 0 else "#60a5fa"
            ctx.fillRect(x * self.cell_size, y * self.cell_size, self.cell_size, self.cell_size)
            
            # Draw border
            ctx.strokeStyle = "#1f2937"
            ctx.strokeRect(x * self.cell_size, y * self.cell_size, self.cell_size, self.cell_size)
        
        # Draw food
        ctx.fillStyle = "#ef4444"
        ctx.beginPath()
        ctx.arc(
            (self.food[0] * self.cell_size) + (self.cell_size / 2),
            (self.food[1] * self.cell_size) + (self.cell_size / 2),
            self.cell_size / 2,
            0,
            2 * 3.14159
        )
        ctx.fill()
        
        # Draw score
        ctx.fillStyle = "#ffffff"
        ctx.font = "20px Arial"
        ctx.textAlign = "left"
        ctx.fillText(f"Score: {self.score}", 10, 30)
        
        # Draw game over message
        if self.game_over:
            ctx.fillStyle = "rgba(0, 0, 0, 0.75)"
            ctx.fillRect(0, 0, self.width, self.height)
            
            ctx.fillStyle = "#ffffff"
            ctx.font = "30px Arial"
            ctx.textAlign = "center"
            ctx.fillText("Game Over", self.width / 2, self.height / 2 - 30)
            ctx.fillText(f"Score: {self.score}", self.width / 2, self.height / 2 + 10)
            ctx.font = "20px Arial"
            ctx.fillText("Press Start to play again", self.width / 2, self.height / 2 + 50)
    
    def reset(self):
        self.snake = [(10, 10), (9, 10), (8, 10)]
        self.direction = "right"
        self.food = self.generate_food()
        self.score = 0
        self.game_over = False

# Function to be called from JavaScript
def init_snake_game(canvas_id):
    canvas = js.document.getElementById(canvas_id)
    ctx = canvas.getContext("2d")
    
    game = SnakeGame(canvas.width, canvas.height, 20)
    
    def game_loop():
        if not game.game_over:
            game.move()
        game.draw(ctx)
    
    def handle_keydown(event):
        key = event.key
        if key == "ArrowUp":
            game.change_direction("up")
        elif key == "ArrowDown":
            game.change_direction("down")
        elif key == "ArrowLeft":
            game.change_direction("left")
        elif key == "ArrowRight":
            game.change_direction("right")
    
    # Set up event listeners
    js.document.addEventListener("keydown", handle_keydown)
    
    # Return game object and loop function for JavaScript to use
    return game, game_loop

File: js/snake.js
document.addEventListener("DOMContentLoaded", () => {
  const canvas = document.getElementById("snake-canvas")
  const ctx = canvas.getContext("2d")
  const startButton = document.getElementById("start-snake")
  const resetButton = document.getElementById("reset-snake")

  const GRID_SIZE = 20
  const CELL_SIZE = canvas.width / GRID_SIZE

  let snake = []
  let food = {}
  let direction = "right"
  let gameRunning = false
  let gameLoop
  let score = 0

  function initGame() {
    snake = [
      { x: 10, y: 10 },
      { x: 9, y: 10 },
      { x: 8, y: 10 },
    ]

    direction = "right"
    score = 0
    generateFood()
    drawGame()
  }

  function generateFood() {
    food = {
      x: Math.floor(Math.random() * GRID_SIZE),
      y: Math.floor(Math.random() * GRID_SIZE),
    }

    // Make sure food doesn't spawn on snake
    for (const cell of snake) {
      if (cell.x === food.x && cell.y === food.y) {
        return generateFood()
      }
    }
  }

  function drawGame() {
    // Clear canvas
    ctx.fillStyle = "#1f2937"
    ctx.fillRect(0, 0, canvas.width, canvas.height)

    // Draw snake
    snake.forEach((cell, index) => {
      ctx.fillStyle = index === 0 ? "#3b82f6" : "#60a5fa"
      ctx.fillRect(cell.x * CELL_SIZE, cell.y * CELL_SIZE, CELL_SIZE, CELL_SIZE)

      // Draw border
      ctx.strokeStyle = "#1f2937"
      ctx.strokeRect(cell.x * CELL_SIZE, cell.y * CELL_SIZE, CELL_SIZE, CELL_SIZE)
    })

    // Draw food
    ctx.fillStyle = "#ef4444"
    ctx.beginPath()
    ctx.arc(food.x * CELL_SIZE + CELL_SIZE / 2, food.y * CELL_SIZE + CELL_SIZE / 2, CELL_SIZE / 2, 0, Math.PI * 2)
    ctx.fill()

    // Draw score
    ctx.fillStyle = "#ffffff"
    ctx.font = "20px Arial"
    ctx.textAlign = "left"
    ctx.fillText(`Score: ${score}`, 10, 30)
  }

  function moveSnake() {
    const head = { x: snake[0].x, y: snake[0].y }

    switch (direction) {
      case "up":
        head.y -= 1
        break
      case "down":
        head.y += 1
        break
      case "left":
        head.x -= 1
        break
      case "right":
        head.x += 1
        break
    }

    // Check for wall collision
    if (head.x < 0 || head.x >= GRID_SIZE || head.y < 0 || head.y >= GRID_SIZE) {
      gameOver()
      return
    }

    // Check for self collision
    for (let i = 0; i < snake.length; i++) {
      if (head.x === snake[i].x && head.y === snake[i].y) {
        gameOver()
        return
      }
    }

    // Add new head
    snake.unshift(head)

    // Check if food eaten
    if (head.x === food.x && head.y === food.y) {
      score += 10
      generateFood()
    } else {
      // Remove tail if no food eaten
      snake.pop()
    }

    drawGame()
  }

  function gameOver() {
    clearInterval(gameLoop)
    gameRunning = false

    ctx.fillStyle = "rgba(0, 0, 0, 0.75)"
    ctx.fillRect(0, 0, canvas.width, canvas.height)

    ctx.fillStyle = "#ffffff"
    ctx.font = "30px Arial"
    ctx.textAlign = "center"
    ctx.fillText("Game Over", canvas.width / 2, canvas.height / 2 - 30)
    ctx.fillText(`Score: ${score}`, canvas.width / 2, canvas.height / 2 + 10)
    ctx.font = "20px Arial"
    ctx.fillText("Press Start to play again", canvas.width / 2, canvas.height / 2 + 50)
  }

  function startGame() {
    if (gameRunning) return

    initGame()
    gameRunning = true
    gameLoop = setInterval(moveSnake, 150)
  }

  function resetGameState() {
    clearInterval(gameLoop)
    gameRunning = false
    initGame()
  }

  // Event listeners
  startButton.addEventListener("click", startGame)
  resetButton.addEventListener("click", resetGameState)

  document.addEventListener("keydown", (e) => {
    if (!gameRunning) return

    switch (e.key) {
      case "ArrowUp":
        if (direction !== "down") direction = "up"
        break
      case "ArrowDown":
        if (direction !== "up") direction = "down"
        break
      case "ArrowLeft":
        if (direction !== "right") direction = "left"
        break
      case "ArrowRight":
        if (direction !== "left") direction = "right"
        break
    }
  })

  // Initialize game
  initGame()
})
